backend/server.js

```diff
@@ -0,0 +1,224 @@
+app.use(cookieParser());
+app.use(cors());
+app.use(cors({ credentials: true, origin: "http://localhost:3000" })); // CONFIG FOR LOCAL SERVER
+app.use("/images", express.static(__dirname + "/uploads")); // Files in local folder uploads have endpoints as /images/x
+app.use("/assets", express.static(__dirname + "/assets"));
+
+//app.use(cors({ credentials: true, origin: "http://134.209.119.133:3000" })); // CONFIG FOR REMOTE SERVER
+
+//_________________Begining of END POINTS____________________
+
+app.get("/", (req, res) => {
+  res.send("hello");
+});
+//----------------------------------------------------------------------------
+
+//SIGNUP
+app.post("/signup", upload.none(), (req, res) => {
+  if (req.body === undefined) {
+    console.log("no data with login request");
+    res.status(400);
+    res.send(
+      JSON.stringify({ success: false, result: "signup data not recieved" })
+    );
+    return;
+  }
+  console.log("login request recieved || user: " + req.body.username);
+  if (!passwordCheck(req.body.password)) {
+    console.log("password failed check");
+    res.status(400);
+    res.send(
+      JSON.stringify({
+        success: false,
+        result:
+          "Password must be 7 or more characters long and contain one lowercase letter, one number, on upper case letter"
+      })
+    );
+    return;
+  }
+  let hashedPassword = passwordHash.generate(req.body.password);
+  Collection_LoginInfo.find({ username: req.body.username }).toArray(
+    (err, foundUsers) => {
+      if (foundUsers.length > 0) {
+        console.log("username taken");
+        res.status(400);
+        res.send(
+          JSON.stringify({ success: false, result: "Username already in use" })
+        );
+        return;
+      }
+      Collection_LoginInfo.insertOne(
+        {
+          username: req.body.username,
+          hashedPassword
+        },
+        (err, result) => {
+          if (err) throw err;
+          console.log(result);
+          console.log("account created");
+          res.status(200);
+          res.send(
+            JSON.stringify({ success: true, result: "account created" })
+          );
+          return;
+        }
+      );
+    }
+  );
+});
+
+//---------------------------------------------------------------------------
+//AutoLogin
+app.post("/auto-login", upload.none(), (req, res) => {
+  let loggedIn = false;
+  console.log(req.cookies.sid);
+  Collection_Sessions.find({ sid: req.cookies.sid }).toArray(
+    (err, foundSessionArr) => {
+      if (foundSessionArr.length < 1) {
+        return;
+      }
+      let foundUser = foundSessionArr[0];
+      console.log("session match");
+      console.log("logged in user " + foundUser.username);
+      loggedIn = true;
+      res.status(200);
+      res.send(
+        JSON.stringify({
+          success: true,
+          result: "logged in as " + foundUser.username
+        })
+      );
+    }
+  );
+});
+//LOGIN
+app.post("/login", upload.none(), (req, res) => {
```

Comment:
For these endpoints, it does feel a bit messy. I think some abstraction could come to the rescue here!

I would create a new backend file called `models.js`. This file would do a few things:

- Open the connection to the DB
- Hold the Collections object
- Export a few nice-to-use methods.

With these new methods, you could imagine this login endpoint looking something like:

```js
app.post("/login", upload.none(), async (req, res) => {
  if (req.body === undefined {
    res.status(400).json(...);
  }

  const foundUser = await lookupUserByUsername(req.body.username);
  
  if (!foundUser) {
    return res.status(401).json(...);
  }

  const isValidPassword = passwordHash.verify(req.body.password, foundUser.hashedPassword);

  if (!isValidPassword) {
    // return an error
  }

  const session = await getOrCreateSession(foundUser)

  res.cookie(sid, session.id)
  res.json({ success: true });
})
```

To be clear: the methods I made up (`lookupUserByUsername`, `getOrCreateSession`) don't exist, you'd have to write them. The idea is to come up with an API that is easy to use in your server.js, so that you can move all the implementation details somewhere else.

When writing your own async methods, you'll likely need to create your own promises. This stuff can get pretty complicated, but essentially you'd do this:

```js
const myNewAsyncMethod = () => {
  return new Promise((resolve, reject) => {
    Collection.loginInfo.find({...}, (err, foundUserArr) => {
      // If you get an error, you can reject the promise
      if (err) {
        return reject(err)
      }

      // do whatever work you want here
      // once you're done:
      return resolve(data);
    })
  });
  });
}
```

Whatever you pass to `resolve()` will get returned:

```js
const justResolve5 = () => new Promise((resolve) => resolve(5));
const someVal = await justResolve5();
// someVal === 5
```

Reply:
I realize this is a humongously complicated comment, haha, but I think it gets at one of the key parts of becoming a better programmer: developing abstractions, coming up with APIs that are pleasant to use. This allows us to write more complicated programs, because we can compartmentalize, and tuck stuff away when we don't need it. When a new developer starts working on this codebase, your abstractions can make the difference between an easy-to-understand onboarding, and a totally inscrutable experience.

So yeah, while this is the trickiest bit of feedback I'm giving, it's probably also the most important. Happy to go over it in more depth!